espresso/streamer_test.go

```diff
@@ -0,0 +1,589 @@
+func (m *MockStreamerSource) AdvanceSafeL2() {
+	m.SafeL2 = createL2BlockRef(m.SafeL2.Number+1, m.FinalizedL1)
+}
+
+// AdvanceEspressoHeightByNBlocks advances the LatestEspHeight by n blocks.
+func (m *MockStreamerSource) AdvanceEspressoHeightByNBlocks(n int) {
+	m.LatestEspHeight += uint64(n)
+}
+
+// AdvanceEspressoHeight advances the LatestEspHeight by one block.
+func (m *MockStreamerSource) AdvanceEspressoHeight() {
+	m.LatestEspHeight++
+}
+
+// SyncStatus returns the current sync status of the mock streamer source.
+// Only the fields FinalizedL1, FinalizedL1, and SafeL2 are populated, as those
+// are the only fields explicitly inspected by the EspressoStreamer.
+func (m *MockStreamerSource) SyncStatus() *eth.SyncStatus {
+	return &eth.SyncStatus{
+		FinalizedL1: m.FinalizedL1,
+		SafeL2:      m.SafeL2,
+	}
+}
+
+func (m *MockStreamerSource) AddEspressoTransactionData(height, namespace uint64, txData esp_client.TransactionsInBlock) {
+	if m.EspTransactionData == nil {
+		m.EspTransactionData = make(map[EspBlockAndNamespace]esp_client.TransactionsInBlock)
+	}
+
+	m.EspTransactionData[BlockAndNamespace(height, namespace)] = txData
+
+	if m.LatestEspHeight < height {
+		m.LatestEspHeight = height
+	}
+}
+
+var _ espresso.L1Client = (*MockStreamerSource)(nil)
+
+// L1 Client methods
+
+func (m *MockStreamerSource) HeaderHashByNumber(ctx context.Context, number *big.Int) (common.Hash, error) {
+	l1Ref := createL1BlockRef(number.Uint64())
+	return l1Ref.Hash, nil
+}
+
+// Espresso Client Methods
+var _ espresso.EspressoClient = (*MockStreamerSource)(nil)
+
+func (m *MockStreamerSource) FetchLatestBlockHeight(ctx context.Context) (uint64, error) {
+	return m.LatestEspHeight, nil
+}
+
+// ErrorNotFound is a custom error type used to indicate that a requested
+// resource was not found.
+type ErrorNotFound struct{}
+
+// Error implements error.
+func (ErrorNotFound) Error() string {
+	return "not found"
+}
+
+// ErrNotFound is an instance of ErrorNotFound that can be used to indicate
+// that a requested resource was not found.
+var ErrNotFound error = ErrorNotFound{}
+
+func (m *MockStreamerSource) FetchTransactionsInBlock(ctx context.Context, blockHeight uint64, namespace uint64) (esp_client.TransactionsInBlock, error) {
+	if m.LatestEspHeight < blockHeight {
+		return esp_client.TransactionsInBlock{}, ErrNotFound
+	}
+
+	// NOTE: if this combination is not found, we will end up returning an
+	//       empty TransactionsInBlock, which is intentional.  It will allow
+	//       the consumer to know that this block exists, but no transactions
+	//       for the requested namespace exist.
+	return m.EspTransactionData[BlockAndNamespace(blockHeight, namespace)], nil
+}
+
+// Espresso Light Client implementation
+var _ espresso.LightClientReaderInterface = (*MockStreamerSource)(nil)
+
+// NoOpLogger is a no-op implementation of the log.Logger interface.
+// It is used to pass a non-nil logger to the EspressoStreamer without
+// producing any output.
```

Comment:
What is the benefit of this NoOpLogger?

Reply:
No direct benefit, other than preventing the spam of logging.  I primarily did it to be able to pass a Logger that adheres to the correct interface, without having to jump through the hoops of figuring out how to create an actual logger.

(We don't do `nil` checks when logging, so this **MUST** not be `nil`)